include/cuda/std/detail/__concepts

```diff
@@ -0,0 +1,282 @@
+  _CUDA_VSTD::requires_<__VA_ARGS__>
+#define _LIBCUDACXX_CONCEPT_FRAGMENT_REQS_REQUIRES_typename(...)               \
+  static_cast<_CUDA_VSTD::_Concept::tag<__VA_ARGS__> *>(nullptr)
+#if defined(__GNUC__) && !defined(__clang__)
+// GCC can't mangle noexcept expressions, so just check that the
+// expression is well-formed.
+// https://gcc.gnu.org/bugzilla/show_bug.cgi?id=70790
+#define _LIBCUDACXX_CONCEPT_FRAGMENT_REQS_REQUIRES_noexcept(...) __VA_ARGS__
+#else
+#define _LIBCUDACXX_CONCEPT_FRAGMENT_REQS_REQUIRES_noexcept(...)               \
+  _CUDA_VSTD::requires_<noexcept(__VA_ARGS__)>
+#endif
+
+#define _LIBCUDACXX_FRAGMENT(NAME, ...)                                        \
+  (1u == sizeof(NAME##_LIBCUDACXX_CONCEPT_FRAGMENT_(                           \
+             static_cast<_CUDA_VSTD::_Concept::tag<__VA_ARGS__> *>(nullptr),   \
+             nullptr)))
+
+#endif
+
+////////////////////////////////////////////////////////////////////////////////
+// _LIBCUDACXX_TEMPLATE
+// Usage:
+//   _LIBCUDACXX_TEMPLATE(typename A, typename B)
+//     (requires Concept1<A> _LIBCUDACXX_AND Concept2<B>)
+//   void foo(A a, B b)
+//   {}
+#if _LIBCUDACXX_CXX_CONCEPTS
+#define _LIBCUDACXX_TEMPLATE(...)                                              \
+  template <__VA_ARGS__> _LIBCUDACXX_PP_EXPAND /**/
+#define _LIBCUDACXX_AND &&                     /**/
+#else
+#define _LIBCUDACXX_TEMPLATE(...)                                              \
+  template <__VA_ARGS__ _LIBCUDACXX_TEMPLATE_SFINAE_AUX_ /**/
+#define _LIBCUDACXX_AND                                                        \
+  &&_LIBCUDACXX_true_, int > = 0, _CUDA_VSTD::enable_if_t < /**/
+#endif
+
+#define _LIBCUDACXX_TEMPLATE_SFINAE(...)                                       \
+  template <__VA_ARGS__ _LIBCUDACXX_TEMPLATE_SFINAE_AUX_ /**/
+#define _LIBCUDACXX_TEMPLATE_SFINAE_AUX_(...)                                  \
+  , bool _LIBCUDACXX_true_ = true,                                             \
+         _CUDA_VSTD::enable_if_t <                                             \
+                 _LIBCUDACXX_PP_CAT(_LIBCUDACXX_TEMPLATE_SFINAE_AUX_3_,        \
+                                    __VA_ARGS__) &&                            \
+             _LIBCUDACXX_true_,                                                \
+         int > = 0 > /**/
+#define _LIBCUDACXX_TEMPLATE_SFINAE_AUX_3_requires
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+namespace _Concept {
+template <typename...> struct tag;
+template <class> inline constexpr bool true_() { return true; }
+} // namespace _Concept
+
+#if defined(__clang__) || defined(_MSC_VER)
+template <bool B> _CUDA_VSTD::enable_if_t<B> requires_() {}
+#else
+template <bool B>
+_LIBCUDACXX_INLINE_VAR constexpr _CUDA_VSTD::enable_if_t<B, int> requires_ = 0;
+#endif
+
+template <class _Ty, class... _Others>
+#if _LIBCUDACXX_STD_VER < 17
+_LIBCUDACXX_CONCEPT _One_of =
+    _CUDA_VSTD::disjunction<_CUDA_VSTD::is_same<_Ty, _Others>...>::value;
```

Comment:
I believe we already backported `disjunction` to work pre C++17 in libcu++.

In general, if there are library constructs like this that we rely on, you should check if we have already backported them, and if not, we can backport it.